python/ray/experimental/sgd/sgd.py

```diff
@@ -0,0 +1,504 @@
+        tf_session_args = {
+            "device_count": {"CPU": num_devices},
+            "log_device_placement": False,
+            "gpu_options": tf.GPUOptions(force_gpu_compatible=True),
+            "inter_op_parallelism_threads": 128,
+        }
+        config_proto = tf.ConfigProto(**tf_session_args)
+        self.sess = tf.Session(config=config_proto)
+        self.models = []
+        grad_ops = []
+
+        if use_cpus:
+            device_tmpl = "/cpu:%d"
+        else:
+            device_tmpl = "/gpu:%d"
+        for device_idx in range(num_devices):
+            device = device_tmpl % device_idx
+            with tf.device(device):
+                with tf.variable_scope("device_%d" % device_idx):
+                    model = model_creator(worker_index, device_idx)
+                    self.models.append(model)
+                    model.grads = [
+                        t for t in model.optimizer.compute_gradients(
+                            model.loss)
+                        if t[0] is not None]
+                    grad_ops.append(model.grads)
+
+        if num_devices == 1:
+           assert not max_bytes, "Not supported with 1 GPU"
+           self.packed_grads_and_vars = grad_ops
+        else:
+            if max_bytes:
+                from tfbench import modified_allreduce
+                self.packed_grads_and_vars, packing_vals = (
+                    modified_allreduce.sum_gradients_all_reduce(
+                        "", grad_ops, 1, all_reduce_alg, 1,
+                        list(range(num_devices)),
+                        agg_small_grads_max_bytes=max_bytes))
+            else:
+                self.packed_grads_and_vars = (
+                    allreduce.sum_gradients_all_reduce(
+                        "", grad_ops, 1, all_reduce_alg, 1,
+                        list(range(num_devices))))
+        self.per_device_grads = [
+            list(zip(*dev_gv))[0] for dev_gv in self.packed_grads_and_vars]
+        assert(len(self.per_device_grads) == num_devices)
+        self.num_grads = num_grads = len(self.packed_grads_and_vars[0])
+        if max_bytes:
+            print("Packed grads => {} tensors".format(num_grads))
+
+        # Ops for reading grads with the right control deps
+        nccl_noops = []
+        for j in range(num_grads)[::-1]:
+            with tf.control_dependencies(
+                    nccl_noops + [dev_grad[j]
+                    for dev_grad in self.per_device_grads]):
+                nccl_noops = [tf.no_op()]
+
+        # You must fetch this otherwise the NCCL allreduce will hang
+        self.nccl_control_out = tf.group(*nccl_noops)
+
+        round_robin_devices = False
+        if plasma_op:
+            store_socket = (
+                ray.worker.global_worker.plasma_client.store_socket_name)
+            manager_socket = (
+                ray.worker.global_worker.plasma_client.manager_socket_name)
+            memcpy_plasma_module = tf.load_op_library(
```

Comment:
This needs to change, see https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_plasma_tf_op.py#L87

Reply:
https://github.com/richardliaw/osdi2018/pull/8/files#diff-625435b68a5254f4ebbca9ccf8d655df